src/Confluent.SchemaRegistry/Rest/RestService.cs

```diff
@@ -0,0 +1,306 @@
+                .ToList();
+        }
+
+        #region Base Requests
+
+        private async Task<HttpResponseMessage> ExecuteOnOneInstanceAsync(HttpRequestMessage request)
+        {
+            // There may be many base urls - roll until one is found that works.
+            //
+            // Start with the last client that was used by this method, which only gets set on 
+            // success, so it's probably going to work.
+            //
+            // Otherwise, try every client until a successful call is made (true even under 
+            // concurrent access).
+
+            string aggregatedErrorMessage = null;
+            HttpResponseMessage response = null;
+            bool gotOkAnswer = false;
+            bool firstError = true;
+            
+            int startClientIndex;
+            lock (lastClientUsedLock)
+            {
+                startClientIndex = this.lastClientUsed;
+            }
+
+            int loopIndex = startClientIndex;
+            int clientIndex = -1; // prevent uninitialized variable compiler error.
+            for (; loopIndex < clients.Count + startClientIndex && !gotOkAnswer; ++loopIndex)
+            {
+                clientIndex = loopIndex % clients.Count;
+
+                try
+                {
+                    response = await clients[clientIndex]
+                            .SendAsync(request).ConfigureAwait(false);
+
+                    // In the case of an internal server error, try another server 
+                    //   (reason could be e.g. "error while forwarding the request to the master")
+                    // 4xx errors should not be retried (these are conclusive)
+                    if (response.StatusCode == HttpStatusCode.InternalServerError)
+                    {
+                        if (!firstError)
+                        {
+                            aggregatedErrorMessage += "; ";
+                        }
+                        else 
+                        {
+                            firstError = false;
+                        }
+
+                        string message = "";
+                        int errorCode = -1;
+                        try
+                        {
+                            var errorObject = JObject.Parse(await response.Content.ReadAsStringAsync().ConfigureAwait(false));
+                            message = errorObject.Value<string>("message");
+                            errorCode = errorObject.Value<int>("error_code");
+                        }
+                        catch
+                        {
+                            aggregatedErrorMessage += $"[{clients[clientIndex].BaseAddress}] {response.StatusCode}";
+                        }
+
+                        aggregatedErrorMessage += $"[{clients[clientIndex].BaseAddress}] {response.StatusCode} {errorCode} {message}";
+                    }
+                    else 
+                    {
+                        gotOkAnswer = true;
```

Comment:
is !500 always ok? What about 4xx?

Reply:
The java client considers them all the same and I'll follow that here too as you suggest. I think my reasoning was that a 4xx error implied a valid response from the server that would be mimicked by all servers. I actually still think that is probably right, though I'm not sure.

Reply:
I retained the separate 4xx functionality in the end when SR responds with valid content, since I think it's potentially useful for the caller to be able to differentiate between the two types of errors. I reworked the logic however, and this should be re-reviewed.